### ISSUE TITLE
node cannot join cluster, but r/o access works

### ISSUE THREAD

**Issue by ibotty** [coreos/etcd#3453]
I am running etcd-2.0.13-2.fc22.x86_64 from fedora atomic host.

node1 has etcd with a populated data dir, node2 has no data dir, but had connected in the past (server failure).

I get the following errors setting up raft. etcdctl r/o access works on both nodes (after node2 got its snapshot).

### node1 log excerpt

```
2015/09/07 09:28:17 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406411
2015/09/07 09:28:17 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:17 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406412
2015/09/07 09:28:17 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:18 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406413
2015/09/07 09:28:18 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:18 etcdhttp: unexpected error: etcdserver: request timed out
2015/09/07 09:28:18 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406414
2015/09/07 09:28:18 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:19 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406415
2015/09/07 09:28:19 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:19 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406416                                                                                 
2015/09/07 09:28:19 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:19 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406417
2015/09/07 09:28:19 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:20 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406418
2015/09/07 09:28:20 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:20 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406419
2015/09/07 09:28:20 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:20 raft: 4f8ea99c760001 received msgApp rejection(lastindex: 9981964) from 4202ea068a2f4e77 for index 10406420
2015/09/07 09:28:20 raft: 4f8ea99c760001 decreased progress of 4202ea068a2f4e77 to [next = 9981965, match = 9981964, wait = 0]
2015/09/07 09:28:21 etcdserver: publish error: etcdserver: request timed out
```

### node2 log excerpt

```
2015/09/07 09:28:14 rafthttp: client streaming to 4f8ea99c760001 at term 660752 has been stopped 
2015/09/07 09:28:14 etcdserver: recovered from incoming snapshot at index 9981964 
2015/09/07 09:28:15 etcdserver: publish error: etcdserver: request timed out 
2015/09/07 09:28:16 raft: 4202ea068a2f4e77 [commit: 9981964] ignored snapshot [index: 9981964, term: 615821] 
2015/09/07 09:28:16 rafthttp: starting client stream to 4f8ea99c760001 at term 660752 
2015/09/07 09:28:16 raft: 4202ea068a2f4e77 [commit: 9981964] ignored snapshot [index: 9981964, term: 615821] 
2015/09/07 09:28:17 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406411] rejected msgApp [logterm: 660752, index: 10406411] from 4f8ea99c760001 
2015/09/07 09:28:17 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406412] rejected msgApp [logterm: 660752, index: 10406412] from 4f8ea99c760001 
2015/09/07 09:28:18 raft: 4202ea068a2f4e77 [commit: 9981964] ignored snapshot [index: 9981964, term: 615821] 
2015/09/07 09:28:18 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406413] rejected msgApp [logterm: 660752, index: 10406413] from 4f8ea99c760001 
2015/09/07 09:28:18 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406414] rejected msgApp [logterm: 660752, index: 10406414] from 4f8ea99c760001 
2015/09/07 09:28:19 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406415] rejected msgApp [logterm: 660752, index: 10406415] from 4f8ea99c760001 
2015/09/07 09:28:19 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406416] rejected msgApp [logterm: 660752, index: 10406416] from 4f8ea99c760001 
2015/09/07 09:28:19 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406417] rejected msgApp [logterm: 660752, index: 10406417] from 4f8ea99c760001 
2015/09/07 09:28:20 raft: 4202ea068a2f4e77 [logterm: 0, index: 10406418] rejected msgApp [logterm: 660752, index: 10406418] from 4f8ea99c760001 
2015/09/07 09:28:20 raft: 4202ea068a2f4e77 [commit: 9981964] ignored snapshot [index: 9981964, term: 615821] 
2015/09/07 09:28:20 etcdserver: publish error: etcdserver: request timed out 
```

### etcd configuration

```
ETCD_NAME=nodeN
ETCD_DATA_DIR="/var/lib/etcd/default.etcd"                                                              
ETCD_LISTEN_CLIENT_URLS="http://IP_OF_NODE_N:2379,http://127.0.0.1:2379"
ETCD_LISTEN_PEER_URLS="http://IP_OF_NODE_N:2380"
ETCD_INITIAL_ADVERTISE_PEER_URLS="http://IP_OF_NODE_N:2380"
ETCD_INITIAL_CLUSTER_STATE="existing"
ETCD_INITIAL_CLUSTER_TOKEN="etcd-cluster"
ETCD_ADVERTISE_CLIENT_URLS="http://IP_OF_NODE_N:2379"
ETCD_DISCOVERY_SRV="my.domain"
ETCD_DISCOVERY_FALLBACK="proxy"
```

### SRV records

are correct afaict.

```
$ dig +noall +answer SRV _etcd-server._tcp.my.domain
_etcd-server._tcp.butter.sh. 3599 INSRV 100 0 2380 node1.my.domain.
_etcd-server._tcp.butter.sh. 3599 INSRV 100 0 2380 node2.my.domain.
$ dig +noall +answer node1.my.domain. node2.my.domain.
node1.my.domain. 19316 INSRV  100 A IP_OF_NODE_1
node2.my.domain. 19312 INSRV  100 A IP_OF_NODE_2
```

Nodes can reach each other on 2379 and 2380 (verified through curl), and there is no selinux alert or anything.

I uploaded full (redacted) logs of [node1](http://paste.fedoraproject.org/264361/41612459) and [node2](http://paste.fedoraproject.org/264362/41612465).

**Comment by yichengq** [comment]
You should not connect a previous member that loses its data dir to the cluster again. This is an unsafe operation and is not allowed.

Could you check the network between members works well? The log says that node2 always cannot send messages to node1.

**Comment by yichengq** [comment]
Check the life cycle of data directory here: https://github.com/coreos/etcd/blob/master/Documentation/admin_guide.md#lifecycle for more details.

**Comment by yichengq** [comment]
Could you check the member list in both nodes? https://github.com/coreos/etcd/blob/master/Documentation/other_apis.md#list-members

**Comment by ibotty** [comment]
Thank you for the feedback.

My main problem seems to be answered in your second comment's link. I'll have to remove the member before adding the new member.

Unfortunately, removing the member did not work, when node2 died. I guessed at that point because of quorum. We will have a three-node cluster in production, I was just testing how etcd would behave when two of three machines die.

Network between both nodes works fine. I started node1's etcd after node2's. That should explain, why the first messages can't be received.


I am closing the issue, as it was clearly me not reading the documentation thoroughly.
